src/client/dfs/xattr.c

```diff
@@ -143,6 +145,8 @@ dfs_getxattr(dfs_t *dfs, dfs_obj_t *obj, const char *name, void *value, daos_siz
 		return EINVAL;
 	if (name == NULL)
 		return EINVAL;
+	if (value == NULL && size != NULL && *size > 0)
+		return EINVAL;
```

Comment:
im actually going to request changes for this. if the buffer itself is NULL, there should be no requirement to set the size value. technically this might break existing apps if we return EINVAL now.

Reply:
@mchaarawi , if the caller does not know how large the EA (xattr) will be, it needs to call getxattr() with \size parameter as zero for detection firstly, then with the returned size value to allocate buffer and re-call getxattr() with the real buffer size. Under such case, for the first call "value == NULL && size != NULL && *size == 0", is it valid use case or not?

Reply:
@Nasf-Fan the "value == NULL && size != NULL && *size == 0" is allowed parameter, but "value == NULL && size != NULL && *size > 0" is not allowed by this PR.

@mchaarawi why don't allow the case of "value == NULL && size != NULL && *size > 0" is below code in this func -
L175.     d_iov_set(&sg_iov, value, *size);
it will generate a sgl for obj_fetch with one iov "iov.iov_buf = NULL, iov.iov_buf_len = *size (>0)", that IOV cannot be a valid and must be reject by client IO.
I don't think this IOV is a valid input parameter.
If you don't want to check that case inside dfs_getxattr(), it will be rejected by client IO parameter check anyway. Another option is for the case of "value == NULL && size != NULL && *size > 0" do "d_iov_set(&sg_iov, NULL, 0)" at L175.

Reply:
> @Nasf-Fan the "value == NULL && size != NULL && *size == 0" is allowed parameter, but "value == NULL && size != NULL && *size > 0" is not allowed by this PR.

If that (value == NULL && size != NULL && *size == 0) is valid use case, then we cannot just check "value == NULL", that is what I want to express.

Reply:
I understand what you mean. Im juut worried about this change in behavior with the additional check:
for object daos_obj_fetch, and daos_kv_get, we explicitly say that if the size of the buffer is unknown, one can set to DAOS_REC_ANY and DAOS will return the buffer size.
In DFS, we do not explicitly say that. The user can think that If the buffer is NULL we would get the size (size[IN] is ignored and need not be initialized). if the buffer is not NULL, we would fill up to size and return the size.  

I can see what you mean and how it can mean an issue for client IO. to make sure we don't introduce regressions, please rerun this with this feature tag:
`Features: dfs dfuse datamover`
thanks.

Reply:
I think that providing a NULL buffer and with an non-zero buffer_size should not be valid parameter for any case.
But I am fine to rerun with "Features: dfs dfuse datamover hdf5" although I don't really think it is necessary TBH.

Reply:
> I think that providing a NULL buffer and with an non-zero buffer_size should not be valid parameter for any case.

the problem is users can forget to initialize the size if just passing a NULL buffer especially in an [in,out] parameter, since the size is to be returned. so i disagree with you there. since this is only done on xattr for DFS, it's not going to be a huge change. we need to update the header file to indicate that. I can take care of this later. 

> But I am fine to rerun with "Features: dfs dfuse datamover hdf5" although I don't really think it is necessary TBH.

you touched DFS code, so Features: dfs is definitely required
you also touched cli_obj.c code in a way that can add errors to users that were not returned before.
dfuse uses xattr a lot, so we need to check dfuse layer if it missed setting size to 0.
hdf5-vol uses the object API and can be affected by this for the same reason.
datamover uses dfuse and dfs heavily and can also make the same usage issue
so you might be introducing CI failures in those tests and can affect future landings. it's better to be safe and detect those to make sure that we fix other places / tests and avoid breaking master for everyone else.